packages/web3-core/src/web3_subscriptions.ts

```diff
@@ -71,14 +72,24 @@ export abstract class Web3Subscription<
 		});
 
 		const messageListener = (
-			err: Error | undefined,
-			data?: JsonRpcSubscriptionResult | JsonRpcNotification<Log>,
+			data?:
+				| JsonRpcSubscriptionResult
+				| JsonRpcSubscriptionResultOld<Log>
+				| JsonRpcNotification<Log>,
 		) => {
-			if (data && jsonRpc.isResponseWithNotification(data)) {
-				this._processSubscriptionResult(data?.params.result);
+			// for old providers
+			if (data?.data) {
+				this._processSubscriptionResult(data?.data?.result ?? data?.data);
+				return;
 			}
-			if (err) {
-				this._processSubscriptionError(err);
+
+			if (
+				data &&
+				jsonRpc.isResponseWithNotification(
+					data as unknown as JsonRpcSubscriptionResult | JsonRpcNotification<Log>,
+				)
+			) {
+				this._processSubscriptionResult(data?.params.result);
```

Comment:
Judging from [this 1.x code](https://github.com/web3/web3.js/blob/13a2edc852aa2ae35cf827f8b08958c26356db4c/packages/web3-core-requestmanager/src/index.js#L109-L118), it seems like this code is to support legacy providers while `data.data.result` (what you have marked as `// for old providers`) is to support EIP-1193

However, to handle the legacy provider, it would appear that we're missing an event listener for `data` event, as `messageListener` is only triggered for `message` events

Reply:
thanks. I've added support for old providers and change comment